### PR TITLE
FIX: layout for mixed descent multiline text objects

### LINKED ISSUE #11468: Layout managers have problems with titles containing MathText
Layout managers seem to have problems with titles containing MathText.

One would expect to have a title which neither overlaps with the axes nor with the figure boundary in all cases depicted below.

### multi line title

tight_layout2

as expected

### multi line title with MathText

tight_layout1

The title is a bit squeezed towards the top figure boundary, but does not go beyond it, nor does it overlap the axes. This is somehow still acceptable.

### single line title with MathText

tight_layout0

Here the title overlaps with the axes. This is in thus far strange because in the case above with multiple lines, this overlap does not occur.

### multi line title

constrained_layout2

as expected

### multi line title with MathText

constrained_layout1

The title exits the figure. This is not desiredeable.

### single line title with MathText

constrained_layout0

Here the title overlaps with the axes. 

## code 

```
import matplotlib
import matplotlib.pyplot as plt
plt.rcParams['savefig.facecolor']='#e9f7fd'

def create_bug(title="", constrained_layout=True, i=0):
    plt.figure(constrained_layout=constrained_layout)
    plt.plot([0,1,2],[0,1,1])
    plt.title(title)
    
    if not constrained_layout:
        plt.tight_layout()
        tx = "tight_layout"
    else:
        tx = "constrained_layout"
    plt.text(0.4,.95, tx, va="center")
    plt.savefig(tx+str(i)+".png")
    
titles = ['line $\sum_{i} x_i$',
          'multi-\nline $\sum_{i} x_i$',
          'multi-\nline']
for layout in [True, False]:
    for i, title in enumerate(titles):
        create_bug(title, layout, i)
```

**Matplotlib version**

  * Operating system: Windows 8.1
  * Matplotlib version: 2.2.2, current master
  * Matplotlib backend : Qt5Agg
  * Python version: 2.7, 3.6

### LINKED ISSUE #11498: Test layout with big descenders and multiple lines inconsistent.
### Bug report

Follow up #11468 

If we have multiline text with big descenders (i.e. `$\sum_i x$`) then the layout mechanism in `Text._get_layout()` doesn't handle the big descenders properly.  Note that `_get_layout()` gives us the bounding box for `text.get_window_extent` so this causes problems for anything that needs to know where the text is and what its size is.  

1. The layout is fine for single lines even large descenders. (first three text objects on each line)
2. For two lines with the same descend the layout is fine (last text object on each line)
3. For mixed size descenders (`test` and ` $\sum_i x$`) the layout is erroneous.  If the big descender comes first, the layout box is too big; conversely if it comes second.  

I keep chasing around solutions, but the algorithm is a bit idiosyncratic so thought I'd post here if anyone had institutional knowledge.  

old

```python
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches

fig, ax = plt.subplots(constrained_layout=False)

ax.set_xlim([0, 1.4])
ax.set_ylim([0, 2])
ax.axhline(0.5, color='C2', linewidth=0.3)
sts = ['Line', '2 Lines g\n 2nd Line g', '$\sum_i x $', 'hi $\sum_i x $\ntest', 'test\n $\sum_i x $', '$\sum_i x $\n $\sum_i x $']

for nn, st in enumerate(sts):
    tt = ax.text(0.2 * nn + 0.1, 0.5, st, horizontalalignment='center',
        verticalalignment='bottom')
    bb = tt.get_window_extent(fig.canvas.get_renderer())
    bbt = bb.inverse_transformed(ax.transAxes)
    r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
    r.set_bounds(bbt.bounds)
    ax.add_patch(r)
ax.text(1.2, 0.5, 'Bottom align', color='C2')


if 1:

    ax.axhline(1.3, color='C2', linewidth=0.3)
    for nn, st in enumerate(sts):
        tt = ax.text(0.2 * nn + 0.1, 1.3, st, horizontalalignment='center',
            verticalalignment='top')
        bb = tt.get_window_extent(fig.canvas.get_renderer())
        bbt = bb.inverse_transformed(ax.transAxes)
        r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
        r.set_bounds(bbt.bounds)
        ax.add_patch(r)
    ax.text(1.2, 1.3, 'Top align', color='C2')


    ax.axhline(1.8, color='C2', linewidth=0.3)
    for nn, st in enumerate(sts):
        tt = ax.text(0.2 * nn + 0.1, 1.8, st, horizontalalignment='center',
            verticalalignment='baseline')
        bb = tt.get_window_extent(fig.canvas.get_renderer())
        bbt = bb.inverse_transformed(ax.transAxes)
        r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
        r.set_bounds(bbt.bounds)
        ax.add_patch(r)
    ax.text(1.2, 1.8, 'Baseline align', color='C2')


    ax.axhline(0.1, color='C2', linewidth=0.3)

    for nn, st in enumerate(sts):
        tt = ax.text(0.2 * nn + 0.1, 0.1, st, horizontalalignment='center',
            verticalalignment='bottom', rotation=20)
        bb = tt.get_window_extent(fig.canvas.get_renderer())
        bbt = bb.inverse_transformed(ax.transAxes)
        r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
        r.set_bounds(bbt.bounds)
        ax.add_patch(r)
    ax.text(1.2, 0.1, 'Bot align, rot20', color='C2')


bb = tt.get_window_extent(fig.canvas.get_renderer())
bbt = bb.inverse_transformed(ax.transAxes)
r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
r.set_bounds(bbt.bounds)
ax.add_patch(r)

fig.savefig('Old.png')

plt.show()
```

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #11498 partially closes #11468

~WIP: this'll fail some tests that depended on the previous bad behaviour...~

EDIT This indeed failed some tests, but I'd pretty strongly argue the tests were broken and this PR makes them correct...

Test: 
```python
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches

fig, ax = plt.subplots(constrained_layout=False)

ax.set_xlim([0, 1.4])
ax.set_ylim([0, 2])
ax.axhline(0.5, color='C2', linewidth=0.3)
sts = ['Line', '2 Lines g\n 2nd Line g', '$\sum_i x $', 'hi $\sum_i x $\ntest', 'test\n $\sum_i x $', '$\sum_i x $\n $\sum_i x $']

for nn, st in enumerate(sts):
    tt = ax.text(0.2 * nn + 0.1, 0.5, st, horizontalalignment='center',
        verticalalignment='bottom')
    bb = tt.get_window_extent(fig.canvas.get_renderer())
    bbt = bb.inverse_transformed(ax.transAxes)
    r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
    r.set_bounds(bbt.bounds)
    ax.add_patch(r)
ax.text(1.2, 0.5, 'Bottom align', color='C2')


if 1:

    ax.axhline(1.3, color='C2', linewidth=0.3)
    for nn, st in enumerate(sts):
        tt = ax.text(0.2 * nn + 0.1, 1.3, st, horizontalalignment='center',
            verticalalignment='top')
        bb = tt.get_window_extent(fig.canvas.get_renderer())
        bbt = bb.inverse_transformed(ax.transAxes)
        r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
        r.set_bounds(bbt.bounds)
        ax.add_patch(r)
    ax.text(1.2, 1.3, 'Top align', color='C2')


    ax.axhline(1.8, color='C2', linewidth=0.3)
    for nn, st in enumerate(sts):
        tt = ax.text(0.2 * nn + 0.1, 1.8, st, horizontalalignment='center',
            verticalalignment='baseline')
        bb = tt.get_window_extent(fig.canvas.get_renderer())
        bbt = bb.inverse_transformed(ax.transAxes)
        r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
        r.set_bounds(bbt.bounds)
        ax.add_patch(r)
    ax.text(1.2, 1.8, 'Baseline align', color='C2')


    ax.axhline(0.1, color='C2', linewidth=0.3)

    for nn, st in enumerate(sts):
        tt = ax.text(0.2 * nn + 0.1, 0.1, st, horizontalalignment='center',
            verticalalignment='bottom', rotation=20)
        bb = tt.get_window_extent(fig.canvas.get_renderer())
        bbt = bb.inverse_transformed(ax.transAxes)
        r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
        r.set_bounds(bbt.bounds)
        ax.add_patch(r)
    ax.text(1.2, 0.1, 'Bot align, rot20', color='C2')


bb = tt.get_window_extent(fig.canvas.get_renderer())
bbt = bb.inverse_transformed(ax.transAxes)
r = mpatches.Rectangle((0,0), 1, 1, clip_on=False, transform=ax.transAxes)
r.set_bounds(bbt.bounds)
ax.add_patch(r)

fig.savefig('Old.png')

plt.show()
```

### Before:

old

### With PR:

new

## PR Checklist

- [x] FIX baseline alignment for multiline
- [ ] Add blue boxes to test...
- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
Squashed, and added boxes around the text in the mulitline test per @efiring suggestion...

File: lib/matplotlib/tests/test_text.py [hunk 1]
```diff
@@ -126,6 +127,55 @@ def test_multiline():
...
+
+    ax.set_xlim([0, 1.4])
+    ax.set_ylim([0, 2])
+    ax.axhline(0.5, color='C2', linewidth=0.3)
+    sts = ['Line', '2 Lineg\n 2 Lg', '$\sum_i x $', 'hi $\sum_i x $\ntest',
+           'test\n $\sum_i x $', '$\sum_i x $\n $\sum_i x $']
+    renderer = fig.canvas.get_renderer()
+
+    def draw_box(ax, tt):
+        if 1:
```
Comment (efiring) [inline lib/matplotlib/tests/test_text.py]: Looks like this conditional is left over from development.

File: lib/matplotlib/text.py [hunk 2]
```diff
@@ -324,14 +324,23 @@ def _get_layout(self, renderer):
...
+                thisy -= max(offsety, (h - d) * self._linespacing)
             horizLayout[i] = thisx, thisy, w, h
             thisy -= d
             width = max(width, w)
             descent = d
 
-        ymin = horizLayout[-1][1]
-        ymax = horizLayout[0][1] + horizLayout[0][3]
+        # Bounding box definition:
+        # by defeinition, ymax is 0
```
Comment (efiring) [inline lib/matplotlib/text.py]: Typo in comment.  Also, a single line of comment should be enough here.

File: lib/matplotlib/text.py [hunk 3]
```diff
@@ -324,14 +324,23 @@ def _get_layout(self, renderer):
...
+        ymax = 0
+        # ymin is the baseline of the last line subtract the descent of the
+        # last line.
```
Comment (efiring) [inline lib/matplotlib/text.py]: I don't understand the comment; can you condense to one line, and clarify?

Comment (jklymak) [reply lib/matplotlib/text.py]: Reworded, though I admit that I don't completely remember what this meant.  This code is very convoluted, and could maybe do with a refactor now that we know what its supposed to do. ...

**Comment by jklymak** [comment]
Pinging for a second review on this...

**Pull request merged**